Assets/Scripts/Characters/EntityController.cs

```diff
@@ -0,0 +1,147 @@
+        this.effectiveStats = new Dictionary<Stat, float> ();
+        this.activeModifiers = new Dictionary<Stat, List<Modifier>> ();
+        foreach (Stat s in System.Enum.GetValues(typeof(Stat))) {
+            baseStats [s] = 0f;
+            effectiveStats [s] = 0f;
+            activeModifiers [s] = new List<Modifier> ();
+        }
+    }
+
+    // Update is called once per frame
+    protected virtual void Update () {
+        // Remove expired Effects and their associated Modifiers
+        for (int i = currentEffects.Count - 1; i >= 0; i--) {
+            var e = currentEffects [i];
+            if (Time.time - e.start_time > e.duration) {
+                activeModifiers [e.modifier.stat].Remove (e.modifier);
+                currentEffects.RemoveAt (i);
+            }
+        }
+        // Calculate effective Stats from base and Modifiers
+        foreach (KeyValuePair<Stat, float> pair in baseStats) {
+            if (activeModifiers.ContainsKey (pair.Key)) {
+                float additive = 0f;
+                float multiplicative = 1f;
+                foreach (Modifier m in activeModifiers[pair.Key]) {
+                    additive += m.additive;
+                    multiplicative += m.multiplicative;
+                }
+                effectiveStats [pair.Key] = (pair.Value + additive) * multiplicative;
+            } else {
+                effectiveStats [pair.Key] = baseStats [pair.Key];
+            }
+        }
+        // Check if dead
+        if (health == 0f) {
+            SetStun (true);
+            // temporary...
+            Quaternion deathRotationQuaternion = Quaternion.LookRotation (new Vector3(0f, -1f, 0f));
+            this.charRigidbody.MoveRotation (deathRotationQuaternion);
+        }
+        if (endStun != 0f) {
+            if (Time.time > endStun) {
+                SetStun (false);
+                endStun = 0f;
+            }
+        }
+        if (endFreeze != 0f) {
+            if (Time.time > endFreeze) {
+                SetFreeze (false);
+                endFreeze = 0f;
+            }
+        }
+
+    }
+
+    public void ApplyEffect(Effect e) {
+        currentEffects.Add (e);
+        activeModifiers [e.modifier.stat].Add (e.modifier);
+    }
+
+    public void TakeDamage(float damage) {
+        health = Mathf.Max (health - damage, 0f);
+    }
+
+    protected void Move (float h, float v) {
+        // Set the movement vector
+        this.movement.Set (h, 0f, v);
+
+        // Normalize the movement vector and make it proportional to the speed per second.
+        this.movement = this.movement.normalized * effectiveStats[Stat.Speed] * Time.deltaTime;
+
+        // Move the character to it's current position plus the movement.
+        this.charRigidbody.MovePosition (this.transform.position + this.movement);
+    }
+
+    /*
+     * Short convenience functions for combination state changes
+     */
+
+    /*
+    * set character cannot turn, move, or act
```

Comment:
Yes. I agree. I wrote the stun, freeze, immobilize stuff before I conceived of effects. I actually made a Trello card to fix this nonsense immediately after pushing this.